### PR TITLE
Lower TableLeftJoinRightDistinct

### PR DESCRIPTION
**Pull request by johnc1231**
Lowers `TableLeftJoinRightDistinct`.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 1]
```diff
@@ -506,6 +506,58 @@ object LowerTableIR {
...
           }
 
+        case TableLeftJoinRightDistinct(left, right, root) =>
+          require(right.typ.keyType isPrefixOf left.typ.keyType)
+
+          val commonKeyLength = right.typ.keyType.size
+          val loweredLeft = lower(left)
+          val leftKeyToRightKeyMap = left.typ.keyType.fieldNames.zip(right.typ.keyType.fieldNames).toMap
+          val newRightPartitioner = loweredLeft.partitioner.strictify.coarsen(commonKeyLength)
+            .rename(leftKeyToRightKeyMap)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:516]: I don't think the rename is necessary. Partitions should ignore field names completely. If that's not true, I'd like to know about it.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 2]
```diff
@@ -506,6 +506,58 @@ object LowerTableIR {
             sorted
           }
 
+        case TableLeftJoinRightDistinct(left, right, root) =>
+          require(right.typ.keyType isPrefixOf left.typ.keyType)
+
+          val commonKeyLength = right.typ.keyType.size
+          val loweredLeft = lower(left)
+          val leftKeyToRightKeyMap = left.typ.keyType.fieldNames.zip(right.typ.keyType.fieldNames).toMap
+          val newRightPartitioner = loweredLeft.partitioner.strictify.coarsen(commonKeyLength)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:515]: You don't want the strictify here. That can make `newRightPartitioner` incompatible with the (unstrictified) left partitioner.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala (hunk 1 above)
Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:516]: I think that the names sort of matter, in that it's pretty common throughout the lowering code to get the key fields by doing `lowered.partitioner.kType`. Let me investigate

Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:516]: Ah, yeah, you're probably right.

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:516]: This is the reason for having this rename:

https://github.com/hail-is/hail/blob/master/hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala#L157

Shuffling uses the key names for the partitioner to `SelectFields`

**Pull request merged**